File: js-Note/generate_ml_pipeline.mjs
import fs from "fs";

// === 配色方案 ===
const COLORS = {
  input: "#66BB6A",
  main: "#2196F3",
  accent: "#FF7043",
  arrow: "#1976D2",
  white: "#FFFFFF",
  bg: "#FFFFFF",
  textDark: "#333333",
  textLight: "#757575",
  noteBg: "#F5F5F5",
  noteBorder: "#BDBDBD",
};

// === 节点数据 (用 SVG 图标替代 emoji) ===
const steps = [
  {
    title: "图像数据采集",
    subtitle: "Image Data Collection",
    details: ["原始人脸图像", "多角度/光照", "数据增强"],
    color: COLORS.input,
    stepNum: "01",
  },
  {
    title: "人脸检测",
    subtitle: "Face Detection",
    details: ["Haar / HOG 特征", "MTCNN / YOLO", "人脸区域裁剪"],
    color: COLORS.main,
    stepNum: "02",
  },
  {
    title: "数据归一化",
    subtitle: "Normalization",
    details: ["灰度转换", "尺寸统一", "像素值归一化"],
    color: COLORS.main,
    stepNum: "03",
  },
  {
    title: "特征提取",
    subtitle: "Feature Extraction",
    details: ["CNN 特征", "LBP / HOG", "降维 PCA"],
    color: COLORS.main,
    stepNum: "04",
  },
  {
    title: "多人脸分类器",
    subtitle: "Multi-Face Classifier",
    details: ["SVM / Softmax", "深度学习分类", "身份识别输出"],
    color: COLORS.accent,
    stepNum: "05",
  },
];

const arrowLabels = ["人脸区域", "裁剪对齐", "特征向量", "分类决策"];

// === SVG 尺寸 ===
const W = 1400;
const H = 420;
const boxW = 210;
const boxH = 250;
const gap = 60;
const arrowW = 70;
const startX = 50;
const boxY = 85;

function escapeXml(s) {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}

// === 生成 SVG ===
let svg = `<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="${W}" height="${H}" viewBox="0 0 ${W} ${H}">
  <defs>
    <filter id="shadow" x="-4%" y="-4%" width="112%" height="112%">
      <feDropShadow dx="2" dy="3" stdDeviation="4" flood-color="#00000018"/>
    </filter>
    <marker id="arrowhead" markerWidth="14" markerHeight="10" refX="14" refY="5" orient="auto">
      <polygon points="0 0, 14 5, 0 10" fill="${COLORS.arrow}"/>
    </marker>
  </defs>

  <!-- 背景 -->
  <rect width="${W}" height="${H}" fill="${COLORS.bg}"/>

  <!-- 标题 -->
  <text x="${W / 2}" y="32" text-anchor="middle" font-family="SimHei, Microsoft YaHei, sans-serif"
        font-size="24" font-weight="bold" fill="${COLORS.textDark}">机器学习多人脸识别流程</text>
  <text x="${W / 2}" y="55" text-anchor="middle" font-family="Arial, sans-serif"
        font-size="12" fill="${COLORS.textLight}" font-style="italic">Multi-Face Recognition Pipeline based on Machine Learning</text>
  <line x1="${W / 2 - 200}" y1="65" x2="${W / 2 + 200}" y2="65" stroke="${COLORS.main}" stroke-width="1.5" opacity="0.3"/>
`;

// === 绘制节点 ===
steps.forEach((step, i) => {
  const x = startX + i * (boxW + gap + arrowW);
  const y = boxY;

  // 阴影层
  svg += `  <rect x="${x + 3}" y="${y + 3}" width="${boxW}" height="${boxH}" rx="12" ry="12"
        fill="#000000" opacity="0.08" filter="url(#shadow)"/>\n`;

  // 主框
  svg += `  <rect x="${x}" y="${y}" width="${boxW}" height="${boxH}" rx="12" ry="12"
        fill="${step.color}" opacity="0.95"/>\n`;

  // 顶部圆角条（装饰）
  svg += `  <rect x="${x}" y="${y}" width="${boxW}" height="6" rx="3" ry="3" fill="white" opacity="0.25"/>\n`;

  // 步骤编号圆形
  svg += `  <circle cx="${x + boxW / 2}" cy="${y + 40}" r="20" fill="white" opacity="0.2"/>\n`;
  svg += `  <text x="${x + boxW / 2}" y="${y + 47}" text-anchor="middle"
        font-family="SimHei, Microsoft YaHei, sans-serif" font-size="20" font-weight="bold" fill="${COLORS.white}">${step.stepNum}</text>\n`;

  // 标题
  svg += `  <text x="${x + boxW / 2}" y="${y + 82}" text-anchor="middle"
        font-family="SimHei, Microsoft YaHei, sans-serif" font-size="16" font-weight="bold" fill="${COLORS.white}">${escapeXml(step.title)}</text>\n`;

  // 英文副标题
  svg += `  <text x="${x + boxW / 2}" y="${y + 102}" text-anchor="middle"
        font-family="Arial, sans-serif" font-size="10" fill="#E3F2FD" font-style="italic">${escapeXml(step.subtitle)}</text>\n`;

  // 分隔线
  svg += `  <line x1="${x + 25}" y1="${y + 118}" x2="${x + boxW - 25}" y2="${y + 118}"
        stroke="white" stroke-opacity="0.25" stroke-width="1"/>\n`;

  // 详细信息
  step.details.forEach((detail, di) => {
    svg += `  <text x="${x + boxW / 2}" y="${y + 142 + di * 26}" text-anchor="middle"
        font-family="SimSun, SimHei, Microsoft YaHei, sans-serif" font-size="12" fill="white" opacity="0.92">- ${escapeXml(detail)}</text>\n`;
  });
});

// === 绘制箭头 ===
for (let i = 0; i < steps.length - 1; i++) {
  const x1 = startX + i * (boxW + gap + arrowW) + boxW + 5;
  const x2 = startX + (i + 1) * (boxW + gap + arrowW) - 5;
  const y = boxY + boxH / 2;
  const midX = (x1 + x2) / 2;

  // 箭头线
  svg += `  <line x1="${x1}" y1="${y}" x2="${x2 - 12}" y2="${y}"
        stroke="${COLORS.arrow}" stroke-width="2.5" marker-end="url(#arrowhead)"/>\n`;

  // 箭头标签
  const labelW = arrowLabels[i].length * 14 + 20;
  svg += `  <rect x="${midX - labelW / 2}" y="${y - 26}" width="${labelW}" height="22" rx="6" ry="6"
        fill="white" stroke="${COLORS.arrow}" stroke-width="0.8"/>\n`;
  svg += `  <text x="${midX}" y="${y - 11}" text-anchor="middle" font-family="SimHei, Microsoft YaHei, sans-serif"
        font-size="11" font-weight="bold" fill="${COLORS.arrow}">${escapeXml(arrowLabels[i])}</text>\n`;
}

// === 底部注释 ===
svg += `  <rect x="${startX}" y="${boxY + boxH + 20}" width="${W - startX * 2}" height="36" rx="8" ry="8"
        fill="${COLORS.noteBg}" stroke="${COLORS.noteBorder}" stroke-width="0.5"/>\n`;
svg += `  <text x="${W / 2}" y="${boxY + boxH + 43}" text-anchor="middle"
        font-family="SimSun, SimHei, Microsoft YaHei, sans-serif" font-size="11" fill="${COLORS.textLight}">关键技术：Haar + MTCNN 人脸检测 → 直方图均衡化归一化 → CNN / PCA 特征提取 → SVM / Softmax 多分类</text>\n`;

svg += `</svg>`;

// === 保存 SVG ===
const svgPath = "ml_pipeline_flowchart.svg";
fs.writeFileSync(svgPath, svg, "utf-8");
console.log(`✅ SVG 已生成: ${svgPath}`);

// === 用 sharp 转 PNG ===
try {
  const sharp = (await import("sharp")).default;
  const pngBuffer = await sharp(Buffer.from(svg)).png({ quality: 100 }).toBuffer();
  fs.writeFileSync("ml_pipeline_flowchart.png", pngBuffer);
  console.log(`✅ PNG 已生成: ml_pipeline_flowchart.png (${(pngBuffer.length / 1024).toFixed(1)} KB)`);
} catch {
  const { execSync } = await import("child_process");
  execSync("npm install sharp", { stdio: "inherit" });
  const sharp = (await import("sharp")).default;
  const pngBuffer = await sharp(Buffer.from(svg)).png({ quality: 100 }).toBuffer();
  fs.writeFileSync("ml_pipeline_flowchart.png", pngBuffer);
  console.log(`✅ PNG 已生成: ml_pipeline_flowchart.png (${(pngBuffer.length / 1024).toFixed(1)} KB)`);
}

File: js-Note/机器学习_/generate_ml_pipeline.py
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from matplotlib.patches import FancyBboxPatch, FancyArrowPatch
import numpy as np

# === 中文字体支持 ===
plt.rcParams.update({
    'font.sans-serif': ['SimHei', 'Microsoft YaHei', 'SimSun', 'Arial Unicode MS'],
    'axes.unicode_minus': False,
    'figure.dpi': 300,
    'savefig.dpi': 300,
})

# === 配色方案 ===
COLOR_BG = '#FFFFFF'
COLOR_BOX = '#2196F3'       # 主蓝色
COLOR_BOX_DARK = '#1565C0'  # 深蓝
COLOR_BOX_LIGHT = '#BBDEFB' # 浅蓝
COLOR_ARROW = '#1976D2'     # 箭头蓝
COLOR_TEXT = '#FFFFFF'       # 白色文字
COLOR_LABEL = '#333333'      # 说明文字
COLOR_ACCENT = '#FF7043'     # 强调色（输出）
COLOR_INPUT = '#66BB6A'      # 输入色

# === 流程节点定义 ===
steps = [
    {
        'title': '图像数据\n采集',
        'subtitle': 'Image Data\nCollection',
        'details': ['原始人脸图像', '多角度/光照', '数据增强'],
        'color': COLOR_INPUT,
        'icon': '📷',
    },
    {
        'title': '人脸检测',
        'subtitle': 'Face Detection',
        'details': ['Haar/HOG特征', 'MTCNN / YOLO', '人脸区域裁剪'],
        'color': COLOR_BOX,
        'icon': '🔍',
    },
    {
        'title': '数据归一化',
        'subtitle': 'Normalization',
        'details': ['灰度转换', '尺寸统一', '像素值归一化'],
        'color': COLOR_BOX,
        'icon': '📐',
    },
    {
        'title': '特征提取',
        'subtitle': 'Feature Extraction',
        'details': ['CNN特征', 'LBP / HOG', '降维PCA'],
        'color': COLOR_BOX,
        'icon': '🧬',
    },
    {
        'title': '多人脸\n分类器',
        'subtitle': 'Multi-Face\nClassifier',
        'details': ['SVM / Softmax', '深度学习分类', '身份识别输出'],
        'color': COLOR_ACCENT,
        'icon': '👤',
    },
]

# === 画布设置 ===
fig, ax = plt.subplots(1, 1, figsize=(14, 4.5))
ax.set_xlim(-0.5, 14.5)
ax.set_ylim(-1.5, 4.5)
ax.set_aspect('equal')
ax.axis('off')
fig.patch.set_facecolor(COLOR_BG)

# === 标题 ===
ax.text(7, 4.1, '机器学习多人脸识别流程', fontsize=16, fontweight='bold',
        ha='center', va='center', color='#212121',
        fontfamily='SimHei')
ax.text(7, 3.55, 'Multi-Face Recognition Pipeline based on Machine Learning',
        fontsize=9, ha='center', va='center', color='#757575', style='italic')

# === 绘制流程节点 ===
box_width = 2.2
box_height = 2.8
start_x = 0.8
gap = 0.7
arrow_len = 0.8

node_positions = []

for i, step in enumerate(steps):
    x = start_x + i * (box_width + gap + arrow_len)
    y = 1.0
    node_positions.append((x, y))

    # 外框阴影
    shadow = FancyBboxPatch(
        (x + 0.06, y - box_height/2 - 0.06),
        box_width, box_height,
        boxstyle="round,pad=0.15",
        facecolor='#E0E0E0', edgecolor='none', alpha=0.5,
        zorder=1
    )
    ax.add_patch(shadow)

    # 主框
    main_box = FancyBboxPatch(
        (x, y - box_height/2),
        box_width, box_height,
        boxstyle="round,pad=0.15",
        facecolor=step['color'], edgecolor='none', alpha=0.92,
        zorder=2
    )
    ax.add_patch(main_box)

    # 顶部图标 + 标题
    ax.text(x + box_width/2, y + box_height/2 - 0.35, step['icon'],
            fontsize=16, ha='center', va='center', zorder=3)
    ax.text(x + box_width/2, y + box_height/2 - 0.75, step['title'],
            fontsize=11, fontweight='bold', ha='center', va='center',
            color=COLOR_TEXT, zorder=3, linespacing=1.3)
    ax.text(x + box_width/2, y + box_height/2 - 1.25, step['subtitle'],
            fontsize=7, ha='center', va='center',
            color='#E3F2FD', zorder=3, style='italic')

    # 分隔线
    sep_y = y + box_height/2 - 1.55
    ax.plot([x + 0.25, x + box_width - 0.25], [sep_y, sep_y],
            color='#FFFFFF', alpha=0.3, linewidth=0.8, zorder=3)

    # 详细信息
    for j, detail in enumerate(step['details']):
        ax.text(x + box_width/2, sep_y - 0.35 - j * 0.38,
                f'• {detail}', fontsize=7.5, ha='center', va='center',
                color=COLOR_TEXT, alpha=0.9, zorder=3)

# === 绘制箭头 ===
for i in range(len(steps) - 1):
    x_start = node_positions[i][0] + box_width + 0.05
    x_end = node_positions[i + 1][0] - 0.05
    y_arrow = 1.0

    # 箭头主体
    arrow = FancyArrowPatch(
        (x_start, y_arrow), (x_end, y_arrow),
        arrowstyle='->,head_width=0.25,head_length=0.15',
        color=COLOR_ARROW, linewidth=2.0,
        connectionstyle='arc3,rad=0',
        zorder=4
    )
    ax.add_patch(arrow)

    # 箭头上的标签
    mid_x = (x_start + x_end) / 2
    arrow_labels = ['检测', '裁剪对齐', '特征向量', '分类决策']
    ax.text(mid_x, y_arrow + 0.32, arrow_labels[i],
            fontsize=7, ha='center', va='center',
            color=COLOR_ARROW, fontweight='bold',
            bbox=dict(boxstyle='round,pad=0.2', facecolor='white',
                     edgecolor=COLOR_ARROW, alpha=0.85, linewidth=0.5),
            zorder=5)

# === 底部注释框 ===
note_y = -0.7
note_box = FancyBboxPatch(
    (0.8, note_y - 0.35), 12.6, 0.7,
    boxstyle="round,pad=0.1",
    facecolor='#F5F5F5', edgecolor='#BDBDBD', linewidth=0.5,
    zorder=1
)
ax.add_patch(note_box)
ax.text(7.1, note_y,
        '关键技术：Haar特征 + MTCNN人脸检测  →  直方图均衡化归一化  →  CNN/PCA特征提取  →  SVM/Softmax多分类',
        fontsize=8, ha='center', va='center', color='#616161', zorder=2)

# === 保存 ===
output_dir = 'D:/GIthub_list/Note/js-Note/'
fig.savefig(output_dir + 'ml_pipeline_flowchart.png', format='png',
            dpi=300, bbox_inches='tight', pad_inches=0.08, facecolor=COLOR_BG)
fig.savefig(output_dir + 'ml_pipeline_flowchart.pdf', format='pdf',
            bbox_inches='tight', pad_inches=0.08, facecolor=COLOR_BG)

print('✅ 图片已生成:')
print(f'   ml_pipeline_flowchart.png (300dpi)')
print(f'   ml_pipeline_flowchart.pdf')

plt.close()
